contracts/sg721-imago/src/contract.rs

```diff
@@ -0,0 +1,448 @@
+
+    // imago
+    let finalizer = deps.api.addr_validate(&msg.finalizer)?;
+    FINALIZER.save(deps.storage, &finalizer)?;
+
+    // sg721 instantiation
+    if msg.collection_info.description.len() > MAX_DESCRIPTION_LENGTH as usize {
+        return Err(ContractError::DescriptionTooLong {});
+    }
+
+    let image = Url::parse(&msg.collection_info.image)?;
+
+
+    if let Some(ref external_link) = msg.collection_info.external_link {
+        Url::parse(external_link)?;
+    }
+
+    let maybe_err = validate_code_uri(msg.code_uri.clone());
+    if maybe_err.is_some() {
+        return Err(maybe_err.unwrap());
+    }
+
+    let royalty_info: Option<RoyaltyInfo> = match msg.collection_info.royalty_info {
+        Some(royalty_info) => Some(RoyaltyInfo {
+            payment_address: deps.api.addr_validate(&royalty_info.payment_address)?,
+            share: royalty_info.share_validate()?,
+        }),
+        None => None,
+    };
+
+    deps.api.addr_validate(&msg.collection_info.creator)?;
+
+    let collection_info = CollectionInfo {
+        creator: msg.collection_info.creator,
+        description: msg.collection_info.description,
+        image: msg.collection_info.image,
+        external_link: msg.collection_info.external_link,
+        royalty_info,
+    };
+
+    COLLECTION_INFO.save(deps.storage, &collection_info)?;
+    CODE_URI.save(deps.storage, &msg.code_uri)?;
+
+    Ok(Response::default()
+        .add_attribute("action", "instantiate")
+        .add_attribute("contract_name", CONTRACT_NAME)
+        .add_attribute("contract_version", CONTRACT_VERSION)
+        .add_attribute("image", image.to_string())
+        .add_messages(fee_msgs))
+}
+
+fn validate_code_uri(uri: String) -> Option<ContractError> {
+    let parsed_token_uri = Url::parse(&uri);
+    if parsed_token_uri.is_err() {
+        return Some(ContractError::InvalidCodeUri {});
+    }
+    if parsed_token_uri.unwrap().scheme() != "ipfs" {
+        return Some(ContractError::InvalidCodeUri {});
+    }
+    return None;
+}
+
+fn finalize_token_uri(deps: DepsMut,
+                      _env: Env,
+                      info: MessageInfo,
+                      token_id: String,
+                      token_uri: String,
+) -> Result<Response, ContractError> {
+    let finalizer = FINALIZER.load(deps.storage)?;
+
+    if info.sender != finalizer {
+        return Err(Unauthorized {});
+    }
+
+    Sg721ImagoContract::default()
+        .tokens
+        .update(deps.storage, &token_id, |token| match token {
+            Some(mut token_info) => {
+                token_info.token_uri = Some(token_uri);
+                token_info.extension = Empty {};
+                Ok(token_info)
+            }
+            None => Err(ContractError::TokenNotFound {
+                got: token_id.to_string(),
+            }),
+        })?;
+
+    return Ok(Response::new()
+        .add_attribute("action", "finalize"));
+}
+
+fn execute_set_code_uri(deps: DepsMut,
```

Comment:
I'm a bit confused for this, doesn't seem to be used by the minter or the contract itself besides for the query, but allows the creator directly to modify this uri, is it needed?

Reply:
This essentially makes the generative art 'on-chain'. Imago reads this parameter off chain and uses it to render the nft preview. For example, when token id 1 is minted, it's included in a block in tx hash `ABC123`. Imago detects a mint event, it reads the `code_uri` parameter from this contract, which is an ipfs url such as `ipfs://def987`. The NFT url is formed by combining the two into this standardized url `ipfs://def987?hash=ABC123`. This hash query parameter is used in the generative work which can be seen in the public works template I've made here [https://github.com/abigpotostew/public-works-template](https://github.com/abigpotostew/public-works-template/blob/4b87d7ba7fa5f74b8bcd23d2d89fa6c20c56122a/src/random.js#L8)

Here's an example from my testnet mints-- `ipfs://QmXYwKAHWJL79C3pEzTcFPDcAfWfqyJwcC81SbpnSju3kS/?hash=5660F2ABED7561BE1131964094B205E481ACF4CF01892AD28C18F402A2499712`


Note that I'm making the code uri mutable in this first version of imago in case there's any rendering issues, creators can update it and re-render the nfts.